File: ts/packages/knowPro/src/answerContextSchema.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { DateRange } from "./interfaces.js";

export type AnswerContext = {
    // Relevant entities
    // Use the 'name' and 'type' properties of entities to PRECISELY identify those that answer the user question.
    entities?: RelevantKnowledge[] | undefined;
    // Relevant topics
    topics?: RelevantKnowledge[] | undefined;
    // Relevant messages
    messages?: RelevantMessage[] | undefined;
};

export type EntityNames = string | string[];

export type RelevantKnowledge = {
    // The actual knowledge
    knowledge: any;
    // Entity or entities who mentioned the knowledge
    origin?: EntityNames | undefined;
    // Entity or entities who received or consumed this knowledge
    audience?: EntityNames | undefined;
    // Time period during which this knowledge was gathered
    timeRange?: DateRange | undefined;
};

export type RelevantMessage = {
    from?: EntityNames | undefined;
    to?: EntityNames | undefined;
    timestamp?: Date | undefined;
    // Text chunks in this message
    messageText?: string | string[] | undefined;
};

File: python/ta/typeagent/knowpro/answer_context_schema.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

# TODO: Are we sure this isn't used as a translator schema class?

from dataclasses import dataclass
from typing import Annotated, Any, Union
from typing_extensions import Doc

from ..knowpro.interfaces import DateRange

EntityNames = Union[str, list[str]]


@dataclass
class RelevantKnowledge:
    knowledge: Annotated[Any, Doc("The actual knowledge")]
    origin: Annotated[
        EntityNames | None, Doc("Entity or entities who mentioned the knowledge")
    ] = None
    audience: Annotated[
        EntityNames | None,
        Doc("Entity or entities who received or consumed this knowledge"),
    ] = None
    timeRange: Annotated[
        DateRange | None, Doc("Time period during which this knowledge was gathered")
    ] = None


@dataclass
class RelevantMessage:
    from_: Annotated[EntityNames | None, Doc("Sender(s) of the message")]
    to: Annotated[EntityNames | None, Doc("Recipient(s) of the message")]
    timestamp: Annotated[str | None, Doc("Timestamp of the message in ISO format")]
    messageText: Annotated[str | list[str] | None, Doc("Text chunks in this message")]


@dataclass
class AnswerContext:
    """Use empty lists for unneeded properties."""

    entities: Annotated[
        list[RelevantKnowledge],
        Doc(
            "Relevant entities. Use the 'name' and 'type' properties of entities to PRECISELY identify those that answer the user question."
        ),
    ]
    topics: Annotated[list[RelevantKnowledge], Doc("Relevant topics")]
    messages: Annotated[list[RelevantMessage], Doc("Relevant messages")]
